File: gaia-sdk-javascript/src/graphql/request/input/BinaryWriteChunkImpulse.ts
import Blob from "cross-blob"

export class BinaryWriteChunkImpulse {
    private readonly uri: string
    private readonly uploadId: string
    private readonly ordinal: number
    private readonly sizeInBytes: number
    private readonly chunk: Blob

    constructor(uri: string, uploadId: string, ordinal: number, sizeInBytes: number, chunk: Blob) {
        this.uri = uri;
        this.uploadId = uploadId;
        this.ordinal = ordinal;
        this.sizeInBytes = sizeInBytes;
        this.chunk = chunk;
    }

    public requestParameters(): any {
        return {
            "uploadId": this.uploadId,
            "ordinal": JSON.stringify(this.ordinal),
            "sizeInBytes": JSON.stringify(this.sizeInBytes),
            "uri": this.uri
        }
    }

    public async data(): Promise<Blob | Buffer> {
        if (typeof FormData !== 'undefined') {
            // browser
            return this.chunk;
        } else {
            // not browser
            // @ts-ignore
            return Buffer.from(await this.chunk.arrayBuffer());
        }
    }
}

File: gaia-sdk-python/gaia_sdk/http/request/BinaryWriteChunkImpulse.py
from typing import Dict
from typing import Any


class BinaryWriteChunkImpulse:
    uri: str
    upload_id: str
    ordinal: int
    size_in_bytes: int
    file: bytes

    def __init__(self, uri: str, upload_id: str, ordinal: int, size_in_bytes: int, file: bytes):
        self.uri = uri
        self.upload_id = upload_id
        self.ordinal = ordinal
        self.size_in_bytes = size_in_bytes
        self.file = file

    def __eq__(self, other):
        return self.uri == other.uri \
               and self.upload_id == other.upload_id \
               and self.ordinal == other.ordinal \
               and self.size_in_bytes == other.size_in_bytes \
               and self.file == other.file

    def __repr__(self):
        return {'uri': self.uri,
                'upload_id': self.upload_id,
                'ordinal': self.ordinal,
                'size_in_bytes': self.size_in_bytes,
                'file': self.file}

    def request_parameters(self) -> Dict[str, Any]:
        return {
            "uri": self.uri,
            "uploadId": self.upload_id,
            "ordinal": self.ordinal,
            "sizeInBytes": self.size_in_bytes
        }

    def data(self) -> bytes:
        return self.file
